crates/goose/src/providers/formats/openai.rs

```diff
@@ -249,24 +542,82 @@ pub fn get_usage(data: &Value) -> Result<Usage, ProviderError> {
+/// * `model_config` - The model configuration to use
+/// * `system` - The system message to include in the request
+/// * `messages` - The messages to include in the request
+/// * `tools` - The tools to include in the request
+/// * `image_format` - The desired format for image content
+///
+/// # Returns
+/// * `anyhow::Result<Value, Error>` - The created request, or an error if the request is invalid
+///
+/// # Example
+/// ```
+/// use goose::model::ModelConfig;
+/// use goose::providers::utils::ImageFormat;
+/// use goose::providers::formats::openai::create_request;
+/// let model_config = ModelConfig {
+///     model_name: "gpt-4".to_string(),
+///     tokenizer_name: "gpt-4".to_string(),
+///     context_limit: Some(4096),
+///     temperature: Some(0.7),
+///     max_tokens: Some(1024),
+/// };
+/// let system = "system";
+/// let messages = vec![];
+/// let tools = vec![];
+/// let image_format = ImageFormat::OpenAi;
+/// let request = create_request(&model_config, system, &messages, &tools, &image_format)?;
+/// assert_eq!(request["model"], "gpt-4");
+/// # Ok::<(), anyhow::Error>(())
+/// ```
 pub fn create_request(
     model_config: &ModelConfig,
     system: &str,
     messages: &[Message],
     tools: &[Tool],
     image_format: &ImageFormat,
 ) -> anyhow::Result<Value, Error> {
-    if model_config.model_name.starts_with("o1-mini") {
+    let is_o1 = model_config.model_name.starts_with("o1");
+    let is_o3 = model_config.model_name.starts_with("o3");
+
+    // Only extract reasoning effort for O1/O3 models
+    let (model_name, reasoning_effort) = if is_o1 || is_o3 {
+        let parts: Vec<&str> = model_config.model_name.split('-').collect();
+        let last_part = parts.last().unwrap();
+
+        match *last_part {
+            "low" | "medium" | "high" => {
+                let base_name = parts[..parts.len() - 1].join("-");
+                (base_name, Some(last_part.to_string()))
+            }
+            _ => (
+                model_config.model_name.to_string(),
+                Some("medium".to_string()),
+            ),
+        }
+    } else {
+        // For non-O family models, use the model name as is and no reasoning effort
+        (model_config.model_name.to_string(), None)
+    };
+
+    if model_name.starts_with("o1-mini") {
         return Err(anyhow!(
             "o1-mini model is not currently supported since Goose uses tool calling and o1-mini does not support it. Please use o1 or o3 models instead."
         ));
     }
 
-    let is_o1 = model_config.model_name.starts_with("o1");
-    let is_o3 = model_config.model_name.starts_with("o3");
-
     let system_message = json!({
-        "role": if is_o1 { "developer" } else { "system" },
+        // NOTE: per OPENAI docs , With O1 and newer models, `developer`
+        // should replace `system` role .
+        // https://platform.openai.com/docs/api-reference/chat/create
+        "role": if is_o1 || is_o3{ "developer" } else { "system" },
```

Comment:
```suggestion
        "role": if is_o1 || is_o3 { "developer" } else { "system" },
```

nit: adding whitespace